Draft: ssh connector: respect StrictHostKeyChecking

bauen1 opened the pull request:
Fixes #737 

In theory this works and has been sort of tested, but I still need to:

- [x] Figure out what unit tests this breaks
- [x] Test that the strict policy actually does what it says
- [x] ~~If necessary split the commit for userknownhostsfile into a seperate PR~~

bauen1 committed:
ssh: don't disable host key checking when performing rsync

This will make ssh behave as configured.

bauen1 committed:
ssh: respect StrictHostKeyChecking

bauen1 committed:
ssh: implement support for userknownhostsfile

themanifold commented:
I did a draft of something similar where I allowed the passing in of custom host keys, but great job.

Fizzadar commented on pyinfra/api/connectors/sshuserclient/client.py:
```diff
@@ -18,6 +19,23 @@
+class AskPolicy(MissingHostKeyPolicy):
+    def missing_host_key(self, client, hostname, key):
+        logger.error('No host key for {0} found in known_hosts'.format(hostname))
+        raise SSHException("AskPolicy: not implemented!\n")
```
This can probably just call `getpass` here the same we do for sudo passwords:

https://github.com/Fizzadar/pyinfra/blob/ff0913d6a172966760b63fe59e55dff9ea852e0d/pyinfra/api/connectors/util.py#L176

bauen1 replied:
This methode is called if paramiko does not find the ssh host key in the users `known_hosts` file, I'm not sure what exactly you mean.

Fizzadar replied:
Sorry I wasn't clear - `getpass` is also the wrong function I meant `raw_input`!

```py
class AskPolicy(MissingHostKeyPolicy):
    def missing_host_key(self, client, hostname, key):
        should_continue = raw_input('No host key for {0} found in known_hosts, continue? [y/N] '.format(hostname))
        if should_continue.lower() == 'y':
                ...add key to known hosts
                return
        raise SSHException('Failed due to no valid host key for {0}'.format(hostname))
```

bauen1 replied:
I still can't find that function

bauen1 replied:
Changed, the code for adding the host key was taken from paramiko itself, but I haven't tested that it actually works.

Fizzadar commented on pyinfra/api/connectors/sshuserclient/client.py:
```diff
@@ -79,6 +106,25 @@ def parse_config(self, hostname, initial_cfg=None, ssh_config_file=None):
+        host_keys_file = None
+        if 'userknownhostsfile' in host_config:
+            host_keys_file = host_config['userknownhostsfile']
+            if host_keys_file.startswith('~'):
+                host_keys_file = getenv('HOME') + host_keys_file[1:]
```
```suggestion
            host_keys_file = os.expanduser(host_keys_file)
```

[`os.expanduser`](https://docs.python.org/3/library/os.path.html#os.path.expanduser) is a neater way to do this.

bauen1 replied:
Done.

Fizzadar reviewed (changes requested):
@bauen1 this is really awesome, a huge improvement to `pyinfra` and very much appreciated.

Left a couple of notes - let me know if there's anything you need a hand with or if you'd like me to sort out the tests.

bauen1 commented:
@themanifold if you want to make your own PR with that It would be quite nice, as my implementation of `UserKnownHostsFile` is incomplete and was only hacked together so pyinfra would work correctly with my setup.

bauen1 commented:
I've tested this, but only the StrictHostKeyPolicy, there is also some potential for additional tests.
Apart from my lack of time to test this thoroughly right now this should be good now.

Fizzadar reviewed (approved):
Updates all look good to me @bauen1, thank you for updating this! When I get a minute I'll write a quick test or two for the other policies so we can get this merged in!

Fizzadar committed:
Stop loading system host keys in SSH connector.

Fizzadar committed:
Ensure host keys are memoized when loading.

This also covers two paramiko issues with host keys, both now have pull
requests on paramiko itself but are handled here.

Most importantly host key files are memoized to avoid excessive load
when executing against many hosts in parallel.

Fizzadar committed:
Fix distro 1.7+ compatibility.

Fizzadar committed:
Update default value for host keys in test.

Fizzadar commented:
Confirmed things working as expected, pushed some unrelated fixes to get the tests working. Thank you once again @bauen1!

Fizzadar committed:
Fix python 2.7 string formatting.

Can't wait to be free of Python 2!
